File: DSAlgorithms/patterns/2_pointers/1_pair_target_sum.ts
const pair_target_sum = (arr: number[], target_sum: number) => {
    let left = 0;
    let right = arr.length - 1;
    while(left < right)
    {
        const current_sum = arr[left] + arr[right];
        if(current_sum == target_sum)
            return [left, right];
        if(target_sum > current_sum)
            left++;
        else 
            right--;
    }
    return [-1, -1];
}
console.log(pair_target_sum([1, 2, 3, 4, 6], 6));
console.log(pair_target_sum([2, 5, 9, 11], 11));

File: DSAlgorithms/patterns/2_pointers/1_pair_target_sum.cpp
#include <bits/stdc++.h>
#include <iostream>
class pair_target_sum {
    public: 
        static std::pair<int, int> search(const std::vector<int> &arr, int target_sum)
        {
            int left = 0;
            int right = arr.size() - 1;
            while(left < right)
            {
                int current_sum = arr[left] + arr[right];
                if(current_sum == target_sum)
                    return std::make_pair(left, right);
                if(target_sum > current_sum)
                    left++;
                else 
                    right--;
            }
            return std::make_pair(-1, -1);
        }
};

int main(int argc, char *argv[])
{
    auto result = pair_target_sum::search(std::vector<int>{1, 2, 3, 4, 6}, 6);
    std::cout<<"Pair with target sum: [" << result.first <<", "<< result.second<< "]"<<'\n';
    result = pair_target_sum::search(std::vector<int>{ 2, 5, 9, 11}, 11);
    std::cout<<"Pair with target sum: [" << result.first <<", "<< result.second<< "]"<<'\n';
}

